renku/api/datasets.py

```diff
@@ -352,69 +356,103 @@ def _add_from_git(self, dataset, path, url, target, **kwargs):
             ])
 
-        src = submodule_path / (target or '')
+        target = target or ''
+        relative_to = relative_to or ''
 
-        if target and relative_to:
+        src = submodule_path / target
+
+        if relative_to:
             relative_to = Path(relative_to)
             if relative_to.is_absolute():
                 assert u.scheme in {
                     '', 'file'
                 }, 'Only relative paths can be used with URLs.'
-                target = (Path(url).resolve().absolute() / target).relative_to(
-                    relative_to.resolve()
+                relative_to = relative_to.resolve().relative_to(
+                    Path(url).resolve().absolute()
                 )
+
+            if target:
+                if str(target) != str(relative_to):
+                    # src already includes target so do not append it
+                    src_dir = src if src.is_dir() else src.parent
+                    target = src_dir.relative_to(submodule_path / relative_to)
             else:
-                # src already includes target so we do not have to append it
-                target = src.relative_to(submodule_path / relative_to)
+                target = relative_to
+
+        if str(target) == '.':
+            target = ''
+        if str(relative_to) == '.':
+            relative_to = ''
 
-        # link the target into the data directory
-        dst = self.path / path / (target or '')
+        submodule_repo = self.repo.submodule(submodule_name).module()
+        src_root = submodule_repo.head.commit.tree
+        if target:
+            src_root = submodule_repo.head.commit.tree / str(target)
 
+        files = [src_root.path]
         # if we have a directory, recurse
         if src.is_dir():
-            files = []
-            dst.mkdir(parents=True, exist_ok=True)
-            # FIXME get all files from submodule index
-            for f in src.iterdir():
-                try:
-                    files.extend(
-                        self._add_from_git(
-                            dataset,
-                            path,
-                            url,
-                            target=f.relative_to(submodule_path),
-                            **kwargs
-                        )
-                    )
-                except ValueError:
-                    pass  # skip files outside the relative path
-            return files
+            files.extend({e.path for e in src_root.traverse()})
+
+        # create a map of all submodule's dataset files to be used to
+        # get creators. Use Dataset's creator if a file does not have
+        # a creator
+        submodule_client = LocalClient(submodule_path)
+        submodule_dataset_creators = {}
+        for sb_dataset in submodule_client.datasets.values():
+            for f in sb_dataset.files:
+                creators = f.creator if f.creator else sb_dataset.creator
```

Comment:
To be consistent with the other default value statements.
```suggestion
                creators = f.creator or sb_dataset.creator
```